compiler/planner/insert_planner.rs

```diff
@@ -0,0 +1,402 @@
+
+#[derive(Debug, Clone, Hash, Ord, PartialOrd, Eq, PartialEq)]
+pub enum VariableSource {
+    TypeSource(TypeSource),
+    ValueSource(ValueSource),
+    ThingSource(ThingSource),
+}
+
+#[derive(Debug, Clone, Hash, Ord, PartialOrd, Eq, PartialEq)]
+pub enum TypeSource {
+    Input(VariablePosition),
+    TypeConstant(usize),
+}
+
+#[derive(Debug, Clone, Hash, Ord, PartialOrd, Eq, PartialEq)]
+pub enum ValueSource {
+    Input(VariablePosition),
+    ValueConstant(usize),
+}
+
+#[derive(Debug, Clone, Hash, Ord, PartialOrd, Eq, PartialEq)]
+pub enum ThingSource {
+    Input(VariablePosition),
+    Inserted(usize),
+}
+
+#[derive(Debug)]
+pub enum InsertInstruction {
+    // TODO: Just replace this with regular `Constraint`s and use a mapped-row?
+    Entity { type_: TypeSource },
+    Attribute { type_: TypeSource, value: ValueSource },
+    Relation { type_: TypeSource },
+    Has { owner: ThingSource, attribute: ThingSource }, // TODO: Ordering
+    RolePlayer { relation: ThingSource, player: ThingSource, role: TypeSource }, // TODO: Ordering
+}
+
+pub struct InsertPlan {
+    pub instructions: Vec<InsertInstruction>,
+    pub type_constants: Vec<answer::Type>,
+    pub value_constants: Vec<Value<'static>>,
+    pub n_created_concepts: usize,
+    pub output_row: Vec<VariableSource>, // Where to copy from
+    pub debug_info: HashMap<VariableSource, Variable>,
+}
+
+pub fn build_insert_plan(
+    input_variables: &HashMap<Variable, VariablePosition>,
+    type_annotations: &TypeAnnotations,
+    constraints: &[Constraint<Variable>],
+) -> Result<InsertPlan, InsertCompilationError> {
+    let mut type_constants = collect_type_from_labels_and_role_names(constraints, type_annotations)?;
+    extend_type_constants_for_untyped_role_players_from_annotations(
+        constraints,
+        type_annotations,
+        &mut type_constants,
+    )?;
+    let value_constants = collect_values_from_expression_and_comparison(constraints, &input_variables)?;
+    let isa_kinds = collect_kind_from_isa(constraints, input_variables, type_annotations)?;
+
+    let mut instructions = Vec::new();
+    filter_variants!(Constraint::Isa : constraints).enumerate().try_for_each(|(i, isa)| {
+        debug_assert!(*isa_kinds.index.get(&isa.thing()).unwrap() == i);
+        let type_ = get_type_source(&input_variables, &type_constants.index, isa.type_())?;
+        let kind = isa_kinds.items[i];
+        let instruction = match kind {
+            Kind::Entity => InsertInstruction::Entity { type_ },
+            Kind::Attribute => {
+                let value = get_value_source(&input_variables, &value_constants.index, isa.thing())?;
+                InsertInstruction::Attribute { type_, value }
+            }
+            Kind::Relation => InsertInstruction::Relation { type_ },
+            Kind::Role => Err(InsertCompilationError::IsaStatementForRoleType { isa: isa.clone() })?,
+        };
+        instructions.push(instruction);
+        Ok(())
+    })?;
```

Comment:
`isa` are executed first so all vertices are available before we move on to the edges.
Even concepts created using `has <type> <value>` will have an `isa` statement.